source/extensions/filters/http/csrf/csrf_filter.cc

```diff
@@ -37,27 +37,39 @@ bool isModifyMethod(const Http::RequestHeaderMap& headers) {
           method_type == method_values.Delete || method_type == method_values.Patch);
 }
 
-absl::string_view hostAndPort(const absl::string_view header) {
-  Http::Utility::Url absolute_url;
-  if (!header.empty()) {
-    if (absolute_url.initialize(header, /*is_connect=*/false)) {
-      return absolute_url.hostAndPort();
+std::string hostAndPort(const absl::string_view absolute_url) {
+  Http::Utility::Url url;
+  if (!absolute_url.empty()) {
+    if (url.initialize(absolute_url, /*is_connect=*/false)) {
+      return std::string(url.hostAndPort());
     }
-    return header;
+    return std::string(absolute_url);
   }
   return EMPTY_STRING;
 }
 
-absl::string_view sourceOriginValue(const Http::RequestHeaderMap& headers) {
-  const absl::string_view origin = hostAndPort(headers.getInlineValue(origin_handle.handle()));
-  if (origin != EMPTY_STRING) {
+// Note: per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin,
+//       the Origin header must include the scheme (and hostAndPort expects
+//       an absolute URL).
+std::string sourceOriginValue(const Http::RequestHeaderMap& headers) {
+  const auto origin = hostAndPort(headers.getInlineValue(origin_handle.handle()));
+  if (!origin.empty()) {
     return origin;
   }
   return hostAndPort(headers.getInlineValue(referer_handle.handle()));
 }
 
-absl::string_view targetOriginValue(const Http::RequestHeaderMap& headers) {
-  return hostAndPort(headers.getHostValue());
+std::string targetOriginValue(const Http::RequestHeaderMap& headers) {
+  const auto host_value = headers.getHostValue();
+
+  // Don't even bother if there's not Host header.
+  if (host_value.empty()) {
+    return EMPTY_STRING;
+  }
+
+  const auto absolute_url = fmt::format(
+      "{}://{}", headers.Scheme() != nullptr ? headers.getSchemeValue() : "http", host_value);
```

Comment:
Is it safe to assume that http1.1 requests can always fallback to http?

Reply:
Not sure -- `headers.Scheme()` is `nullptr` during integration tests, but shouldn't it always be there?

Reply:
`:scheme` is an http2 pseudo-header so it wouldn't be attached to any requests using http1.1. My concern about making this assumption is that requests coming from an origin using https without a scheme header would now fail.

Reply:
Ah right. Hmm. Actually, don't infer the scheme from requests in other places? Let me look around.

Reply:
I take it back -- actually doing the fallback to http is fine. Because, the `Url::hostAndPort()` method will drop the port when it's a known port (80 or 443). So in practice, mixing http and https is fine. I did add a test for that case.

Reply:
awesome! looks good to me then